File: app/src/main/java/jp/wildtree/android/apps/hhsadvrev/ZRuleBase.java
package jp.wildtree.android.apps.hhsadvrev;

public class ZRuleBase {
	
	private final int _map_id;
	private final int _cmd_id;
	private final int _obj_id;
    private final ZRuleBlock[] _rules;

	public static final int file_block_size = 96;
	public static final int EndOfRule = 0xff;
	private static final int RULE_BLOCK_LENGTH = (file_block_size / 4 - 1);
	
	public ZRuleBase(byte[] b)
	{
		_rules  = new ZRuleBlock[RULE_BLOCK_LENGTH];
		
		_map_id = (b[0] & 0xff);
		_cmd_id = (b[1] & 0xff);
		_obj_id = (b[2] & 0xff);
		for (int i = 0 ; i < RULE_BLOCK_LENGTH ; i++)
		{
			byte[] rb = new byte[4];
			System.arraycopy(b, 4 + 4 * i, rb, 0, 4);
			_rules[i] = new ZRuleBlock(rb);
		}
	}
	
	public boolean endOfRule()
	{
		return (_map_id == 0xff);
	}
	
	public int mapId()
	{
		return _map_id;
	}
	
	public int cmdId()
	{
		return _cmd_id;
	}
	
	public int objId()
	{
		return _obj_id;
	}
	
	public boolean about(int mapId, int cmdId, int objId)
	{
		if (mapId == _map_id || _map_id == 0)
		{
			if (cmdId == _cmd_id || _cmd_id == 0)
			{
				if (objId == _obj_id || _obj_id == 0)
				{
					return true;
				}
			}
		}
		return false;
	}
	
	public boolean about(ZSystemParams p)
	{
		return about(p.mapId(), p.cmdId(), p.objId());
	}
	
	public boolean condBlock(ZSystemParams p, ZUserData u)
	{
		boolean ok = true;
		
		
		
		return ok;
	}
	
	public boolean run(MainActivity main)
	{
		if (about(main.zSystem))
		{
			boolean cond_ok = true;
			boolean act_ok = true;
			int i = 0;
			while (_rules[i].actCmp())
			{
				cond_ok = _rules[i++].doCompare(main.zSystem, main.userData);
				if (!cond_ok)
				{
					return false; // cond fail.
				}
			}
			if (cond_ok) // all clear. go action phase
			{
				while (_rules[i].op() != ZRuleBlock.ACT_NOP)
				{
					act_ok = _rules[i++].doAction(main) && act_ok;
				}
				if (act_ok)
				{
					main.msgByResId(R.string.msg_okay);
				}
			}
			return cond_ok;
		}
		return false;
	}
}

File: app/src/main/java/jp/wildtree/android/apps/hhsadvrev/ZRuleBase.kt
package jp.wildtree.android.apps.hhsadvrev


class ZRuleBase(b: ByteArray) {
    private val mapId: Int = (b[0].toInt() and 0xff)
    private val cmdId: Int = (b[1].toInt() and 0xff)
    private val objId: Int = (b[2].toInt() and 0xff)
    private val rules: Array<ZRuleBlock?> = arrayOfNulls(RULE_BLOCK_LENGTH)

    init {
        for (i in 0 until RULE_BLOCK_LENGTH) {
            val rb = ByteArray(4)
            System.arraycopy(b, 4 + 4 * i, rb, 0, 4)
            rules[i] = ZRuleBlock(rb)
        }
    }

    fun endOfRule(): Boolean {
        return (mapId == 0xff)
    }

    fun about(m: Int, c: Int, o: Int): Boolean {
        if (m == mapId || mapId == 0) {
            if (c == cmdId || cmdId == 0) {
                if (o == objId || objId == 0) {
                    return true
                }
            }
        }
        return false
    }

    fun about(p: ZSystemParams): Boolean {
        return about(p.mapId(), p.cmdId(), p.objId())
    }

    fun run(main: MainActivity): Boolean {
        main.zSystem?.let {
            if (about(it)) {
                var condOk = true
                var actOk = true
                var i = 0
                while (rules[i]!!.actCmp()) {
                    main.userData?.let { u -> condOk = rules[i++]!!.doCompare(main.zSystem!!, u) }
                    if (!condOk) {
                        return false // cond fail.
                    }
                }
                while (rules[i]!!.op != ZRuleBlock.ACT_NOP) {
                    actOk = rules[i++]!!.doAction(main) && actOk
                }
                if (actOk) {
                    main.msgByResId(R.string.msg_okay)
                }
                return true
            }
        }
        return false
    }

    companion object {
        const val FILE_BLOCK_SIZE: Int = 96
        const val END_OF_RULE: Int = 0xff
        private const val RULE_BLOCK_LENGTH = (FILE_BLOCK_SIZE / 4 - 1)
    }
}
